### ISSUE TITLE
Unnecessary Parentheses does not report violation on assignment within conditional

### ISSUE THREAD

**Issue by shashwat70** [checkstyle/checkstyle#8239]
Check documentation: https://checkstyle.org/config_coding.html#UnnecessaryParentheses

```
C:\Users\Shashwat\Documents>cat config.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name="Checker">
  <module name="TreeWalker">
    <module name="UnnecessaryParentheses">
    </module>
  </module>
</module>

C:\Users\Shashwat\Documents>cat Test.java
public class Main
{
        public static void main(String[] args) {
                int n=2;
                if(((n=n%2))==0) // warning expected
                System.out.println("yes");
        }
}

C:\Users\Shashwat\Documents>java -jar checkstyle-8.31-all.jar -c config.xml Test.java
Starting audit...
Audit done.
```

---------------

There is a redundant set of parentheses on line 5 which should have been reported.

**Comment by pbludov** [comment]
Same for
```
public class Test
{
    int x = ((1 + 2)) * 3;
}
```

the AST:
```
ASSIGN -> = [3:10]
 `--EXPR -> EXPR [3:22]
     `--STAR -> * [3:22]
         |--LPAREN -> ( [3:12]
         |--LPAREN -> ( [3:13]
         |--PLUS -> + [3:16]
         |   |--NUM_INT -> 1 [3:14]
         |   `--NUM_INT -> 2 [3:18]
         |--RPAREN -> ) [3:19]
         |--RPAREN -> ) [3:20]
         `--NUM_INT -> 3 [3:24]
```

The check should be updated to detect nested parentheses.
May be we need a new check for this, but a see no reason to allow nested parentheses and forbid unnecessary parentheses or vice versa.

**Comment by pbludov** [comment]
Another missing violation:
```java
public class TestClass {

    public TestClass() {
        (foo()).bar();            // Expected violation here
        (new TestClass()).bar();  // Expected violation here
    }

    private void bar() {
    }

    private TestClass foo() {
        return this;
    }

}
```
